envoy/upstream/host_description.h

```diff
@@ -55,6 +55,48 @@ struct HostStats {
   }
 };
 
+/**
+ * Weakly-named load metrics to be reported as EndpointLoadMetricStats. Individual stats are
+ * accumulated by calling add(), which combines stats with the same name. The aggregated stats are
+ * retrieved by calling latch(), which also clears the current load metrics.
+ */
+class LoadMetricStats {
+public:
+  struct Stat {
+    uint64_t num_requests_with_metric = 0;
+    double total_metric_value = 0.0;
+  };
+
+  using StatsMap = absl::flat_hash_map<std::string, Stat>;
+
+  LoadMetricStats() : map_(std::make_unique<StatsMap>()) {}
+
+  // Adds the given stat to the map. If the stat already exists in the map, then the stat is
+  // combined with the existing map entry by incrementing num_requests_with_metric and summing the
+  // total_metric_value fields.
+  void add(const std::string& key, double value) {
+    absl::MutexLock lock(&mu_);
+    Stat& stat = (*map_)[key];
+    ++stat.num_requests_with_metric;
+    stat.total_metric_value += value;
+  }
+
+  // Returns an owning pointer to the current load metrics and clears the map.
+  std::unique_ptr<StatsMap> latch() {
+    absl::MutexLock lock(&mu_);
+    std::unique_ptr<StatsMap> latched = std::move(map_);
+    map_ = std::make_unique<StatsMap>();
+    return latched;
+  }
+
+private:
+  absl::Mutex mu_;
+  std::unique_ptr<StatsMap> map_ ABSL_GUARDED_BY(mu_);
```

Comment:
IIUC this adds a per-host map, even in the case where no extension is using `LoadMetricStats`. This adds to the mem usage, which might be significant for use-cases where many endpoints are needed.

Reply:
I feel like I mentioned this in internal comms but can we have a name->index "descriptor map" which is shared by all hosts, and in each host just have a vector of stats accessed via descriptor index?

Reply:
@jmarantz I guess the question there would be the potential thread safety issues in updating the descriptor map. Do we then have to lock a resource that's shared across all hosts?

Reply:
I think you indicated this before but I had forgotten: the set of names you might want to use for these keys is not known at compile-time, because it comes in headers, so it needs to be dynamically changed? That's the issue?

Maybe you can add comments describing how a string becomes a key into this map.

Reply:
Also ISTM that for a given deployment we  could potentially know exactly what the names are, and could lock them in on startup.

Having a string-map for every host with all the strings repeated seems pretty expensive.

Reply:
Correct, the set of names is not known at compile-time. Part of the intent here is to able to utilize this API to implement the Open RCA spec (https://github.com/envoyproxy/envoy/issues/6614) wherein we can arbitrarily define additional metrics. It seems to me that ability runs counter to having the set of keys/metric names pre-defined in a deployment.

Reply:
I expanded the comment to `add()`. It should also be mentioned - the map gets cleared when `latch()` is called. So the keys, if they are present in the map, can be known to be in current use. Once an extension (or the host) stops reporting a metric, that metric will stop taking up memory.